packages/runtime/src/state/actions.ts

```diff
@@ -25,9 +27,12 @@ export const ActionTypes = {
 
   REGISTER_DOCUMENT: 'REGISTER_DOCUMENT',
   UNREGISTER_DOCUMENT: 'UNREGISTER_DOCUMENT',
-
   CHANGE_DOCUMENT: 'CHANGE_DOCUMENT',
 
+  CREATE_ELEMENT_TREE: 'CREATE_ELEMENT_TREE',
+  DELETE_ELEMENT_TREE: 'DELETE_ELEMENT_TREE',
+  CHANGE_ELEMENT_TREE: 'CHANGE_ELEMENT_TREE',
+
```

Comment:
Part of the fix involves maintaining "flattened" element trees in the state; new actions to support this.